tests/test_websocket_list.py

```diff
@@ -0,0 +1,33 @@
+from unittest.mock import MagicMock
+
+import pytest
+
+from websocket_list import WebSocketList
+
+
+@pytest.mark.asyncio
+async def test_get_ordered_by_ping_returns_single_server(backend_client):
+    cell_id = 0
+    address = "cm2-waw1.cm.teststeam.com:27039"
+    backend_client.get_servers.return_value = [address]
+    ssl_context = MagicMock()
+
+    websocket_list = WebSocketList(backend_client, ssl_context)
+
+    servers = await websocket_list.get(cell_id)
+
+    assert len(servers) == 1
+    assert servers[0] == f"wss://{address}/cmsocket/"
+
+
+@pytest.mark.asyncio
+async def test_get_ordered_by_ping_queries_list_with_given_cell_id(backend_client):
+    cell_id = 1
+    address = "cm2-waw1.cm.teststeam.com:27039"
+    backend_client.get_servers.return_value = [address]
+    ssl_context = MagicMock()
+    websocket_list = WebSocketList(backend_client, ssl_context)
+
+    await websocket_list.get(cell_id)
+
+    backend_client.get_servers.assert_called_once_with(cell_id)
```

Comment:
What about `backend_client.side_effect = BackendError()`? Is it layer above supposed to handle that? (I think so bu t don't remember)

What if that would happen on second `get` in a row? 
Maybe add test case that tests using of the in-memory cache. (`backend_client.get_servers.assert_not_called()`)

Reply:
`BackendError` is handled by layer above (`websocket_client`). I don't think `backend_client.get_servers()` will throw it though. I think all errors that occur in `backend_client.get_servers()` will bubble up right to `plugin`. As it did before.

As for in-memory cache, I'm not sure we understand it the same way. Basically it's just an in-memory variable the `websocket_client_ensure_connected()` iterates through.

The rough structure is:
```
websocket_client
  def run()
    while True:
      try:
        self._ensure_connected():
      except: ....

  def _ensure_connected():
    while True:
    for server in servers
      try:
        connect()
      except: ....
```

So there are two `while True` loops with various exception handling. If the `_ensure_connected()` loop fails, the external loop would cause `get()` to be called again. Which is probably desired.

Also `websocket_client` already has some test for error handling, like `test_servers_cache_retry`.

But overall I don't feel very confident changing large parts of `websocket_client` without some global acceptance tests. And this change is intended to help people with `servers_cache` not working anymore and forcing plugin into 'Offline'.

Reply:
> As for in-memory cache, I'm not sure we understand it the same way. Basically it's just an in-memory variable the `websocket_client_ensure_connected()` iterates through.

You're right, I've forgotten that there are 2 for loops, hence difference in understanding usage of in-memory cache. To sum up:
- your solution: each time we lost connection (first loop) we fetch fresh list of servers and iterate over them starting from the first one steam backend has returned
- my previous proposal: each time we lost connection, we try to connect with next server in the servers cache (that is populated only once). Technically it could be done by using generator outside `_ensure_conneted`.

My solution is better in hypocritical case where first steam server on the list has worse quality than subsequent servers. Your solution is much better in case user has switched location (in real or VPN) or plugin was running for a longer time.

So let us stick with your solution.

>  I don't think `backend_client.get_servers()` will throw it though

BackendError may be raised by `AuthenticatedHttpClient` by https://github.com/gogcom/galaxy-integrations-python-api/blob/master/src/galaxy/http.py#L113

It is handled perfectly fine, in case of any troubles, we wait `RECONNECT_INTERVAL_SECONDS` and try again.

Reply:
Just adding comment - we could create some solution between mine and yours, like:
- fetch for new servers list at most once per 10min
- 
but maybe do not optimize prematurely.